### PR TITLE
Add alphanum support

### PR DESCRIPTION
**Pull request by s-tlk**
This will add alphanum support

### CODE REVIEW

File: src/click/alphanum.c [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+#include <fcntl.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "letmecreate/click/alphanum.h"
+#include "letmecreate/click/thermo3.h"
```
Comment (francois-berder-imgtec) [inline src/click/alphanum.c]: Are you sure you need to include thermo3 header ?

File: src/click/alphanum.c [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+#include "letmecreate/core/common.h"
+#include "letmecreate/core/gpio.h"
+#include "letmecreate/core/spi.h"
+#include "letmecreate/core/i2c.h"
+
+static uint8_t current_mikrobus_index = MIKROBUS_2;
+
+uint8_t gpio_pin_le2;
+uint8_t gpio_pin_oe;
+uint8_t gpio_pin_oe2;
```
Comment (francois-berder-imgtec) [inline src/click/alphanum.c]: These variables should be static

File: src/click/alphanum.c [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+			gpio_pin_le2 = MIKROBUS_2_RST;
+			gpio_pin_oe = MIKROBUS_2_INT;
+			gpio_pin_oe2 = MIKROBUS_2_PWM;
+			break;
+		default:
+			printf("Error: Bus not found!\n");
+			return -1;
+	}
+
+	gpio_init(gpio_pin_le2);
```
Comment (francois-berder-imgtec) [inline src/click/alphanum.c]: check the return value of gpio_init and gpio_set_direction.

File: src/click/alphanum.c [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+	gpio_set_direction(gpio_pin_oe, GPIO_OUTPUT);
+	gpio_set_direction(gpio_pin_oe2, GPIO_OUTPUT);
+
+	/* SPI int */
+	if ((ret = spi_init()) != 0) {
+		printf("Error spi init failed with %i\n", ret);
+		return ret;
+	}
+
+	spi_select_bus(bus);
```
Comment (francois-berder-imgtec) [inline src/click/alphanum.c]: Always let the user select the right bus. Wrappers should never change that otherwise it is going to be very confusing for the user to know which bus is currently selected.

File: src/click/alphanum.c [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+			++i;
+		}
+	}
+}
+
+/* 
+ * Used to get the temperature from the temp clicker
+ */
+void
+alphanum_get_temperature(char *str)
```
Comment (francois-berder-imgtec) [inline src/click/alphanum.c]: Why is this here ?

File: include/letmecreate/click/alphanum.h [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+int alphanum_write(uint16_t, uint16_t);
+int alphanum_init(uint8_t);
+void alphanum_switch_cycles(int);
+void alphanum_write_running_text(const char *, int);
+void alphanum_get_temperature(char *);
+
+/*
+ * Char table
+ */
+const uint16_t char_table[51]={
```
Comment (francois-berder-imgtec) [inline include/letmecreate/click/alphanum.h]: This should be moved to alphanum.c and should be static.

File: include/letmecreate/click/alphanum.h [hunk 7]
```diff
@@ -0,0 +1,71 @@
+
+#define SWITCH_INTERVAL 8 * 1000
+#define STR_MAX 10
```
Comment (francois-berder-imgtec) [inline include/letmecreate/click/alphanum.h]: A similar macro is already defined in core/common.h

File: include/letmecreate/click/alphanum.h [hunk 8]
```diff
@@ -0,0 +1,71 @@
+
+#define SWITCH_INTERVAL 8 * 1000
```
Comment (francois-berder-imgtec) [inline include/letmecreate/click/alphanum.h]: This should be moved to alphanum.c

File: include/letmecreate/click/alphanum.h [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+#define STR_MAX 10
+
+uint16_t alphanum_get_char(char);
+int alphanum_write(uint16_t, uint16_t);
+int alphanum_init(uint8_t);
+void alphanum_switch_cycles(int);
+void alphanum_write_running_text(const char *, int);
+void alphanum_get_temperature(char *);
```
Comment (francois-berder-imgtec) [inline include/letmecreate/click/alphanum.h]: It would be great if you could add some documentation for each function and also add a name for each argument.

File: include/letmecreate/click.h [hunk 10]
```diff
@@ -25,5 +25,6 @@
 #include "letmecreate/click/relay2.h"
 #include "letmecreate/click/thermo3.h"
 #include "letmecreate/click/relay.h"
+#include "letmecreate/click/alphanum.h"
```
Comment (francois-berder-imgtec) [inline include/letmecreate/click.h]: I try to respect alphabetic order here, so could you add it just after alcohol.h. Also, this commit should be squashed with the previous commit

**Comment by francois-berder-imgtec** [comment]
I also forgot to say that you need to add header guards in alphanum.h. And the example won't be compiled unless you create an executable in examples/CMakeLists.txt.

File: src/click/alphanum.c (hunk 1 above)
Comment (s-tlk) [inline src/click/alphanum.c]: Yes, it is for use with the -t switch to display the temperature from the thermo3 clicker.

**Comment by s-tlk** [comment]
I implemented most of the requested changes, but it is difficult to test without cmake on the creator board.
Do you plan to build a package for this?

File: src/click/alphanum.c (hunk 1 above)
Comment (francois-berder) [inline src/click/alphanum.c]: I do not see the purpose of having thermo3 related code in the alphanum wrapper. In your example, you could add a function that would do that.

**Comment by francois-berder** [comment]
Do not use usleep, as it is obsolete (although still present). You also seem to have conflicts. Also, did you mention the Alphanum click in Readme.md ?

File: examples/alphanum/main.c [hunk 11]
```diff
@@ -20,8 +18,15 @@
...
+	int ret = 0;
 	char str[] = "Hi"; /* String to print */
 
+	/* SPI int */
+	if ((ret = spi_init()) != 0) {
+		printf("Error spi init failed with %i\n", ret);
+		return ret;
+	}
+
 	/* Initialize the alphanum clicker */
```
Comment (francois-berder) [inline examples/alphanum/main.c]: if you are using mikrobus 2, then you have to call spi_select_bus before.

**Comment by francois-berder-imgtec** [comment]
You need to split your commits into three pull request: one for the alphanum, one for the fix in co.h and one for the oled display.

**Comment by francois-berder** [comment]
Closing pull request. Wrappers for OLED and Alphanum are now part of LetMeCreate in dev branch
